File: mergesort/mergesort.py
def mergesort(list_a, list_b):
	out = []
	i = 0
	j = 0

	while True:

		if list_a[i] < list_b[j]:
			out.append(list_a[i])
			i+=1
		else:
			out.append(list_b[j])
			j+=1

		if i == len(list_a):
			out += list_b[j:]
			break
		if j == len(list_b):
			out += list_a[i:]
			break

	return out


def sort(my_list):
	n = len(my_list)

	if n <= 1:
		return my_list

	return mergesort(sort(my_list[:n/2]), sort(my_list[n/2:]))

print sort([3,5,4,1,2,7,8,998,5,4,2,11,45,4,77])

File: mergesort/mergesort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int mergesort(int listA[], int sizeA, int listB[], int sizeB, int *buffer){
	int i = 0;
	int j = 0;

	while (1) {
		if (listA[i] < listB[j]) {
			*(buffer+i+j) = listA[i];
			i++;
		} else {
			*(buffer+i+j) = listB[j];
			j++;
		}

		if (i == sizeA) {
			while (j < sizeB) {
				*(buffer+i+j) = listB[j];
				j++;
			}
			return 0;
		}
		if (j == sizeB) {
			while (i < sizeA) {
				*(buffer+i+j) = listA[i];
				i++;
			}
			return 0;
		}
	}
}

int sort(int* list, int size, int* buffer) {
	if (size == 1) {
		*buffer = *list;
		return 0;
	}

	int a[size/2];
	int b[size - size/2];

	sort(list, size/2, a);
	sort(list + size/2, size - size/2, b);

	mergesort(a, size/2, b, size - size/2, buffer);
	return 0;
}

int main(){

	int SIZE = 20;
	int list[SIZE];
	int sorted[SIZE];

	int i;
	time_t t;

	srand((unsigned) time(&t));

	for (i = 0; i < SIZE; i++) {
		list[i] = rand()%1000;
	}

	sort(list, SIZE, sorted);

	for (i = 0; i < SIZE; i++) {
		printf("%d ", list[i]);
	}
	printf("\n");

	for (i = 0; i < SIZE; i++) {
		printf("%d ", sorted[i]);
	}
	printf("\n");

	return 0;
}